test/test_python_dispatch.py

```diff
@@ -137,9 +137,23 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
         )
         self.assertExpectedRaisesInline(
             RuntimeError, lambda: A(torch.zeros(1)).detach(),
-            """detach returned invalid type str, expected Tensor"""
+            """Unable to cast Python instance of type <class 'str'> to C++ type 'at::Tensor'"""
```

Comment:
This error message is not as nice as it used to be... Why is that?

Reply:
Previously, this error message came from shallow_copy_and_detach, which calls concrete_detach_fn, which hardcodes the error message:
https://github.com/pytorch/pytorch/blob/f5a71ec2d6956a1ba393657f4b8297bb931eeec2/torch/csrc/autograd/python_variable.cpp#L1828

After this PR, the first detach that gets called is not one from shallow_copy_and_detach. This one does not go through `concrete_detach_fn` so it goes through the usual pybind error reporting mechanism and pybind is very undescriptive (see the error message that is two lines up from this one)